src/tagmanager/tm_ctags.c

```diff
@@ -271,12 +279,149 @@ void tm_ctags_clear_ignore_symbols(void)
 }
 
 
+/* call after all tags have been collected so we don't have to handle reparses
+ * with the counter (which gets complicated when also subparsers are involved) */
+static void rename_anon_tags(TMSourceFile *source_file)
+{
+	GHashTable *anon_counter_table = g_hash_table_new_full(g_str_hash,
+		g_str_equal, g_free, NULL);
+	GPtrArray *removed_typedefs = g_ptr_array_new();
+	guint i;
+
+	for (i = 0; i < source_file->tags_array->len; i++)
+	{
+		TMTag *tag = TM_TAG(source_file->tags_array->pdata[i]);
+		if (tag->flags & tm_tag_flag_anon_t)
+		{
+			gchar *orig_name, *new_name = NULL;
+			guint j;
+			guint new_name_len, orig_name_len;
+			guint scope_len = tag->scope ? strlen(tag->scope) : 0;
+			gchar kind = tag->flags >> 16;
+			const gchar *kind_name = tm_ctags_get_kind_name(kind, tag->lang);
+
+			/* clear the kind temporarily stored in flags in init_tag() */
+			tag->flags &= tm_tag_flag_max_t;
+
+			orig_name = tag->name;
+			orig_name_len = strlen(orig_name);
+
+			if (source_file->lang == TM_PARSER_C || source_file->lang == TM_PARSER_CPP)
+			{
+				/* First check if there's a typedef behind the scope nesting
+				 * such as typedef struct {} Foo; - in this case we can replace
+				 * the anon tag with Foo */
+				for (j = i + 1; j < source_file->tags_array->len; j++)
+				{
+					TMTag *nested_tag = TM_TAG(source_file->tags_array->pdata[j]);
+					guint nested_scope_len = nested_tag->scope ? strlen(nested_tag->scope) : 0;
+
+					/* Nested tags have longer scope than the parent - once the scope
+					 * is equal or lower than the parent scope, we are outside the tag's
+					 * scope. */
+					if (nested_scope_len <= scope_len)
+						break;
+				}
+
+				/* We are out of the nesting - the next tag could be a typedef */
+				if (j < source_file->tags_array->len)
+				{
+					TMTag *typedef_tag = TM_TAG(source_file->tags_array->pdata[j]);
+					guint typedef_scope_len = typedef_tag->scope ? strlen(typedef_tag->scope) : 0;
+
+					/* Should be at the same scope level as the anon tag */
+					if (typedef_tag->type == tm_tag_typedef_t &&
+						typedef_scope_len == scope_len &&
+						g_strcmp0(typedef_tag->var_type, tag->name) == 0)
+					{
+						/* set the name of the original anon tag and pretend
+						 * it wasn't a anon tag */
+						tag->name = g_strdup(typedef_tag->name);
+						tag->flags &= ~tm_tag_flag_anon_t;
+						new_name = tag->name;
+						/* the typedef tag will be removed */
+						g_ptr_array_add(removed_typedefs, GUINT_TO_POINTER(j));
+					}
+				}
+			}
+
+			/* there's no typedef name for the anon tag so let's generate one  */
+			if (!new_name)
+			{
+				gchar buf[50];
+				guint anon_counter = GPOINTER_TO_UINT(g_hash_table_lookup(anon_counter_table, kind_name));
```

Comment:
I think a hash_table is a overweight here. Since `kind` is in the range 0...255 a simple array would sufficie

guint anon_counter = anon_counter_table[kind]++;

Alternatively, you could use a hash table that uses `kind` as the key and not a more expensive hash of `kind_name`

Reply:
Good point, will change. I was also thinking I could allocate the `removed_typedefs` GPtrArray lazily if needed so it isn't allocated every time (and unnecessarily for most languages).